core/src/subgraph/registrar.rs

```diff
@@ -668,10 +679,79 @@ fn create_subgraph_version(
+        .map(|data_source| {
+            let source_address = data_source.source.address;
+            let block_number = match data_source
+                .source
+                .start_block
+                .and_then(|block_str| block_str.parse::<u64>().ok())
+            {
+                Some(number) => number,
+                None => {
+                    warn!(
+                        logger,
+                        "Start block invalid or not specified for the '{}' data source",
+                        &data_source.name
+                    );
+                    return None
+                },
+            };
+
+            // Validate that the start block provided for the Data Source is indeed the block that
+            // contains the contract creation transaction. Allow the block to be used as the
+            // subgraph level start block regardless, but emit a warning.
+            match ethereum_adapter
+                .validate_start_block(logger, block_number, source_address)
+                .wait()
+            {
+                Ok((block, validated)) => {
+                    if !validated {
+                        warn!(
+                            logger,
+                            "Start block for '{}' may lead to skipping on chain data, block number: {}, address: {}",
+                            &data_source.name,
+                            &block_number.to_string(),
+                            source_address
+                                .map(|a| a.to_string())
+                                .unwrap_or("unspecified".to_string())
+                        );
+                    }
+                    Some(block)
+                },
+                Err(e) => {
+                    warn!(
+                        logger,
+                        "Unable to validate start block for '{}', block_num: {}, error: {}",
+                        &block_number.to_string(),
+                        &data_source.name,
+                        e
+                    );
+                    None
+                }
+            }
+        })
+        .filter(|&n| n.is_some())
+        .collect();
+
     // Create deployment only if it does not exist already
     if !deployment_exists {
         let chain_head_block = chain_store.chain_head_ptr()?;
-        let genesis_block = chain_store.genesis_block_ptr()?;
+        let start_blocks = start_blocks.unwrap();
+        let genesis_block = if start_blocks.is_empty() {
+            chain_store.genesis_block_ptr()?
+        } else {
+            EthereumBlockPointer::earliest_block(start_blocks)
+                .or_else(|| chain_store.genesis_block_ptr().ok())
+                .unwrap()
+        };
+        info!(
+            logger,
+            "Set the earliest available Ethereum block, start_block: {}", genesis_block.number
+        );
         ops.extend(
             SubgraphDeploymentEntity::new(&manifest, false, false, genesis_block, chain_head_block)
```

Comment:
Have you tested this to work? I don't get how the start block here gets picked up by the block stream.

Reply:
Ya, I've tested it successfully. 
Store.block_ptr() picks up the `latestEthereumBlock` and uses it to reconcile the next step.   It's tough to grep the codebase for it because of the switching between camel and snake case between our database naming and rust field naming.

Reply:
Ah I see now, thanks. But I believe this will actually start from `genesis_block + 1`, because the subgraph ptr is assumed to have been processed. This is really a more fundamental bug because it means we never process the genesis block. I suppose that's hard to fix, but it's important to work around this for the start block.